Add algorithms that describe how type/state are assigned/updated, and how they influence audio sessions and audio focus.

youennf opened the pull request.

youennf commented:
This PR is a bit bigger and introduces the algorithms themselves. It requires careful review.

alastor0325 commented on index.bs:
```diff
@@ -175,6 +175,88 @@ partial interface Navigator {
+    1. If |audioSession|.[=AudioSession/[[type]]=] is the same as |audioSession|.[=AudioSession/[[appliedType]]=], abort these steps.
+    1. Set |audioSession|.[=AudioSession/[[appliedType]]=] to |audioSession|.[=AudioSession/[[type]]=].
+    1. [=Update all AudioSession states=] of |audioSession|'s [=top-level browsing context=] with |audioSession|.
+    1. For each |element| of |audioSession|.[=AudioSession/[[elements]]=], [=update an element|update=] |element|.
```
nits : `update and element` sounds vague. What does `update` mean? Do you want to refer `to run the steps of 'Update the selected audio session'? If so, it would be worth to mention the full name of the step and add a link for it.

youennf replied:
`update an element` is defined in the audio sink/source section which is filled in the other PR. See https://pr-preview.s3.amazonaws.com/youennf/audio-session/pull/34.html#update-an-element

The principle is that, when the audio session changes, the elements of the audio session might need to be updated.

alastor0325 commented on index.bs:
```diff
@@ -175,6 +175,88 @@ partial interface Navigator {
+    1. Set |audioSession|.[=AudioSession/[[appliedType]]=] to |audioSession|.[=AudioSession/[[type]]=].
+    1. [=Update all AudioSession states=] of |audioSession|'s [=top-level browsing context=] with |audioSession|.
+    1. For each |element| of |audioSession|.[=AudioSession/[[elements]]=], [=update an element|update=] |element|.
+    1. Let |newType| be the result of [=compute the audio session type|computing the type=] of |audioSession|.
```
Could you explain why we need to compute audio session type? The `compute the audio session type` algorithm seems being used to only set the type to a session if a session is active. Any specific reason we want to prohibit setting type on a non-active audio session?

In addition, in `compute the audio session type` algorithm, why it doesn't cover all the types?

youennf replied:
When type is `auto`, we could just set the audio session type to `auto`.
When activating the audio session as `auto`, it is good to be more precise and state for instance that an `auto` audio session with audible media elements is actually a `playback` audio session.

It is indeed not covering all types, I should add transient solo, and maybe make `ambient` the default type.

alastor0325 replied:
Sounds good, let's add more types into the section of `compute the audio session type` later. Thanks!

alastor0325 commented on index.bs:
```diff
@@ -175,6 +175,88 @@ partial interface Navigator {
+
+## Update AudioSession's state ## {#audio-session-update-state-algorithm}
+
+When an audio session [=audio session/element=] is starting or stopping, the user agent will run steps that <dfn>change the state</dfn> of an [=audio session=].
```
We would probably need another PR to define `start and stop` for different audio session element. We also need to define when this algorithm will really need to be run.

Eg. if a AudioSession has several different audio session elements, only one of them stops, but other are still playing, in this case, should this algorithm be triggered?

youennf replied:
This is done in https://github.com/w3c/audio-session/pull/34.
See https://pr-preview.s3.amazonaws.com/youennf/audio-session/pull/34.html#integration

> if a AudioSession has several different audio session elements, only one of them stops, but other are still playing, in this case, should this algorithm be triggered?

The UA should run steps to decide whether to inactivate the audio session (and decide to leave the audio session active).
The algorithm in https://pr-preview.s3.amazonaws.com/youennf/audio-session/pull/34.html#integration) needs to be revised.

alastor0325 replied:
Sounds good, let's do the modification in #34.

alastor0325 commented on index.bs:
```diff
@@ -175,6 +175,88 @@ partial interface Navigator {
+* It can [=inactivate=] {{AudioSession}} objects of the [=top-level browsing context=], as defined in the algorithms below.
+* It can pause the audio of another tab or another application.
+
+Conversely, an [=audio session=] [=audio session/state=] can be modified outside of audio session [=audio session/element=] changes, for instance in case an active [=audio session=] is interrupted.
```
nits : it will probably be good if we can add some example items here to explicitly describe the possible scenarios?

youennf replied:
What kind of example do you have in mind?
There is https://w3c.github.io/audio-session/#interrutpion-handling-example.
I'll make the current wording more precise.

alastor0325 replied:
I was thinking about providing more general examples by using [the example section](https://pr-preview.s3.amazonaws.com/youennf/audio-session/pull/34.html#w3c-conventions), not coding examples.

One of the scenarios you already mentioned is `interrupted by phone call`, and we could provide more examples of interruption, and put them all in the example section for the clearer view.

eg.
```
<div class=example>
For instance, an active `playback` audio session can be interrupted by an incoming phone call, or by another `playback` session that is going to start playing a new media content (... or other examples)
</div>
```

youennf replied:
OK, will add this.

alastor0325 commented on index.bs:
```diff
@@ -175,6 +175,88 @@ partial interface Navigator {
+To <dfn>inactivate</dfn> an {{AudioSession}} named |audioSession|, the user agent MUST run the following steps:
+1. If |audioSession|.[=AudioSession/[[state]]=] is {{AudioSessionState/inactive}}, abort these steps.
+1. Run the following steps [=in parallel=]:
+    1. [=Change the state=] of |audioSession|'s [=audio session=] to {{AudioSessionState/inactive}}.
```
When I click `Change the state`, it directs me to `5.2. Update AudioSession’s state`. If so, for the consistency, should we say `Update the state` instead?

youennf replied:
I am trying to distinguish two things:
- web page does something and UA updates the audio session state based on it (inactivating or activating it).
- the audio session stage changes outside of web page activity, and UA updates the AudioSession state.

I used `update the state` for AudioSession and `change the state` for audio session.
Any idea how to make that clearer?

youennf replied:
Clicking on `change the state` should point to the change state definition.
It is true it is close to the update the state.

alastor0325 replied:
TBF simply reading the spec, it's hard to distinguish those two things.

> When an audio session [element](https://pr-preview.s3.amazonaws.com/youennf/audio-session/pull/33.html#audio-session-element) is starting or stopping, the user agent will run steps that change the state of an [audio session](https://pr-preview.s3.amazonaws.com/youennf/audio-session/pull/33.html#audio-session).

This paragraph doesn't even mention the state change is caused by outside of web page activity. If it's important to let readers understand the differences, maybe we could add your explanation into the spec as well?

youennf replied:
I'll change to `set the state` and `notify the state's change` wording, hoping this is clearer.

alastor0325 commented on index.bs:
```diff
@@ -175,6 +175,88 @@ partial interface Navigator {
+To <dfn>try activating</dfn> an {{AudioSession}} named |audioSession|, the user agent MUST run the following steps:
+1. If |audioSession|.[=AudioSession/[[state]]=] is {{AudioSessionState/active}}, abort these steps.
+1. Run the following steps [=in parallel=]:
+    1. [=Change the state=] of |audioSession|'s [=audio session=] to {{AudioSessionState/active}}. [=Change the state|Changing the state=] to {{AudioSessionState/active}} can fail, in which case the [=audio session=]'s [=audio session/state=] will either be {{AudioSessionState/inactive}} or {{AudioSessionState/interrupted}}.
```
Ditto, should we say `Update the state`? In addition, could we add an example here to demonstrate the situation of the failure of updating state?

youennf replied:
Ditto.

I'll add an example of potential failure (starting a video playback during a phone call for instance).

alastor0325 commented on index.bs:
```diff
@@ -175,6 +175,88 @@ partial interface Navigator {
+To <dfn>update the selected audio session</dfn> of a [=top-level browsing context=] named |context|, the user agent MUST run the following steps:
+1. Let |activeAudioSessions| be the list of all the [=audio session|audio sessions=] [=tied to=] {{AudioSession}} objects of |context| and its children in a breadth-first order, that match both constraints:
+    1. Its [=audio session/state=] is {{AudioSessionState/active}}.
+    1. Its [=audio session/type=] is an [=exclusive type=].
```
Not directly related with this PR, but I wonder shouldn't `transient-solo` be considered an exclusive type as well?

youennf replied:
Indeed, will fix this.

alastor0325 commented on index.bs:
```diff
@@ -175,6 +175,88 @@ partial interface Navigator {
+    1. Its [=audio session/type=] is an [=exclusive type=].
+1. If |activeAudioSessions| is empty, abort these steps.
+1. If there is only one [=audio session=] in |activeAudioSessions|, set the [=selected audio session=] to this [=audio session=] and abort these steps.
+1. Assert that for any {{AudioSession}} object [=tied to=] an [=audio session=] in |activeAudioSessions|'s named |audioSession|, |audioSession|.[=AudioSession/[[type]]=] is {{AudioSessionType/auto}}.
```
Could you explain why we need to assert the type being `auto`? Why can't those audio sessions be other types?

youennf replied:
The idea is that there can be at most only one active audio session with a type that is explicitly set to an exclusive type, since they interrupt each other.
But `auto` audio sessions do not interrupt each other (that would not be web compatible).
So, if we have more than one session in |activeAudioSessions|, it means that they should all be `auto`.

alastor0325 replied:
I think it's worth to add your explanation as a `note` section in the spec. (`<div class="note">`)

youennf replied:
Will add it.

youennf replied:
Done

alastor0325 replied:
nits : it would be great to add `NOTE` explicitly, like what other specs are doing. Eg. [EME](https://w3c.github.io/encrypted-media/#introduction), [MSE](https://w3c.github.io/media-source/#dfn-initialization-segment).

youennf replied:
@tidoust, do you know how to fix this?
The goal would be to render `<div class=note>` with a big green `NOTE`.

youennf commented:
@alastor0325, thanks for the review, I addressed some points and will do a change in the other PR.
Let me know your thoughts on the updates and questions.

alastor0325 commented:
@youennf I apologize for the late reply. I’ve added a few comments, mostly minor suggestions, but overall it looks great to me. Thank you!

youennf commented:
Thanks @alastor0325, I updated the PR according your comments. PTAL.

youennf commented:
Gentle ping

alastor0325 reviewed (approved):
Apologies for the delay on this review, and thank you so much for your help!

chrisn commented on index.bs:
```diff
@@ -175,6 +175,101 @@ partial interface Navigator {
+1. The user agent MAY apply specific heuristics to reorder |activeAudioSessions|.
+1. Set the [=selected audio session=] to the first [=audio session=] in |activeAudioSessions|.
+
+## Other algorithms ## {#audio-session-other-algorithms}
```
A minor suggestion, we could have a section per algorithm, rather than grouping them under "other"?

youennf replied:
The assertion failure means that the UA implementation (or the spec) is wrong and should be fixed.

youennf replied:
As of splitting the other section, let's do that in a follow-up.
